### PR TITLE
Commit pack

### PR DESCRIPTION
**Pull request by Nekotekina**

### CODE REVIEW

File: rpcs3/Emu/RSX/GL/GLGSRender.cpp [hunk 1]
```diff
@@ -1106,6 +1106,15 @@ void GLGSRender::onexit_thread()
...
 
+	if (m_scale_offset_buffer)
+		m_scale_offset_buffer.remove();
+
+	if (m_vertex_constants_buffer)
+		m_vertex_constants_buffer.remove();
+
+	if (m_fragment_constants_buffer)
+		m_fragment_constants_buffer.remove();
+
```
Comment (vlj) [inline rpcs3/Emu/RSX/GL/GLGSRender.cpp]: didn't know the buffer weren't clean in object destructor, thanks for the fix.

File: rpcs3/Emu/RSX/GSManager.cpp [hunk 2]
```diff
@@ -22,23 +22,19 @@ void GSInfo::Init()
...
 	switch (rpcs3::state.config.rsx.renderer.value())
 	{
 	default:
-	case rsx_renderer_type::Null : m_render = new NullGSRender(); break;
-	case rsx_renderer_type::OpenGL: m_render = new GLGSRender(); break;
+	case rsx_renderer_type::Null : m_render = std::make_shared<NullGSRender>(); break;
+	case rsx_renderer_type::OpenGL: m_render = std::make_shared<GLGSRender>(); break;
 #ifdef _MSC_VER
-	case rsx_renderer_type::DX12: m_render = new D3D12GSRender(); break;
+	case rsx_renderer_type::DX12: m_render = std::make_shared<D3D12GSRender>(); break;
```
Comment (vlj) [inline rpcs3/Emu/RSX/GSManager.cpp]: Why not std::unique_ptr ? Is the pointer shared in other place too ?

File: Utilities/SharedMutex.h [hunk 3]
```diff
@@ -1,46 +1,136 @@
...
 
-#include <shared_mutex>
-
-// An attempt to create lock-free (in optimistic case) implementation similar to std::shared_mutex;
-// MSVC implementation of std::shared_timed_mutex is not lock-free and thus may be slow, and std::shared_mutex is not available.
-class shared_mutex_t
+//! An attempt to create effective implementation of "shared mutex", lock-free in optimistic case.
+//! All locking and unlocking may be done by single LOCK XADD or LOCK CMPXCHG instructions.
+//! MSVC implementation of std::shared_timed_mutex seems suboptimal.
+//! std::shared_mutex is not available until C++17.
```
Comment (vlj) [inline Utilities/SharedMutex.h]: Is this shared_mutex API compatible with the std one ? ie Does std::unique_lock/ std::lock_guard compatible with it ?
I think it should be precised in this comment.

Comment (Nekotekina) [inline Utilities/SharedMutex.h]: It's "Mutex" concept, std::unique_lock or std::lock_guard just call lock() and unlock() methods appropriately. I think it's pretty obvious.

**Pull request merged**